odl/space/ntuples.py

```diff
@@ -1767,55 +1939,103 @@ def inner(self, x1, x2):
         inner : float or complex
             The inner product of the two provided vectors
         """
-        inner = _inner_default(x1 * self.vector, x2)
-        if is_real_dtype(x1.dtype):
-            return float(inner)
+        if self.exponent != 2.0:
+            raise NotImplementedError('No inner product defined for '
+                                      'exponent != 2 (got {}).'
+                                      ''.format(self.exponent))
+        else:
+            inner = _inner_default(x1 * self.vector, x2)
+            if is_real_dtype(x1.dtype):
+                return float(inner)
+            else:
+                return complex(inner)
+
+    def norm(self, x):
+        """Calculate the vector-weighted norm of a vector.
+
+        Parameters
+        ----------
+        x : `Fn.Vector`
+            Vector whose norm is calculated
+
+        Returns
+        -------
+        norm : float
+            The norm of the provided vector
+        """
+        if self.exponent == 2.0:
+            return super().norm(x)
```

Comment:
Ah you mean that we optimize the base class `dist` and `norm`? Okay, in that case you're right, although I don't see how to do that with enough generality to not break future implementations of `Fn`-like spaces.



> I would in this case assume that `super().norm()` gave me the "standard R^n norm", not some magic depending on how inner is implemented in this function.

Okay, admittedly this usage is a bit abuse of `super()` in the way that I assume that `inner()` is called. In general, however, your expectations of `super()` are not exactly in line with reality. Despite the name, it's not referring to a "superclass", but that's another topic. I make an explicit call here instead.